packages/core/src/browser/frontend-connection-status.ts

```diff
@@ -0,0 +1,278 @@
+/*
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { inject, injectable, optional } from 'inversify';
+import { Endpoint } from './endpoint';
+import { ILogger } from '../common/logger';
+import { AbstractDialog } from './dialogs';
+import { Event, Emitter } from '../common/event';
+import { MessageService } from '../common/message-service';
+import { StatusBar, StatusBarAlignment } from './status-bar/status-bar';
+import { FrontendApplicationContribution, DefaultFrontendApplicationContribution } from './frontend-application';
+
+/**
+ * Service for listening on backend connection changes.
+ */
+export const ConnectionStatusService = Symbol('ConnectionStatusService');
+export interface ConnectionStatusService {
+    readonly onStatusChange: Event<ConnectionStatusChangeEvent>;
+}
+
+/**
+ * Connection status change event.
+ */
+export interface ConnectionStatusChangeEvent {
+
+    /**
+     * The current state of the connection.
+     */
+    readonly state: ConnectionState,
+
+    /**
+     * Optional health, percentage number.
+     */
+    readonly health?: number
+
+}
+
+/**
+ * The connection-status states.
+ */
+export enum ConnectionState {
+    ONLINE,
+    OFFLINE
+}
+
+@injectable()
+export class ConnectionStatusOptions {
+
+    static DEFAULT: ConnectionStatusOptions = {
+        requestTimeout: 1000,
+        retry: 5,
+        retryInterval: 2000,
+    };
+
+    /**
+     * Timeout for the HTTP GET request in milliseconds.
+     */
+    readonly requestTimeout: number;
+
+    /**
+     * Number of accepted timeouts.
+     */
+    readonly retry: number;
+
+    /**
+     * Retry interval in milliseconds.
+     */
+    readonly retryInterval: number;
+
+}
+
+@injectable()
+export class FrontendConnectionStatusService implements ConnectionStatusService, FrontendApplicationContribution {
+
+    protected readonly statusChangeEmitter: Emitter<ConnectionStatusChangeEvent>;
+    private readonly endpointUrl: string;
+
+    private connectionState: ConnectionStateMachine;
+    private timer: number | undefined;
+
+    constructor(
+        @inject(ConnectionStatusOptions) @optional() protected readonly options: ConnectionStatusOptions = ConnectionStatusOptions.DEFAULT,
+    ) {
+        this.statusChangeEmitter = new Emitter<ConnectionStatusChangeEvent>();
+        this.endpointUrl = new Endpoint().getRestUrl().toString();
+        this.connectionState = new ConnectionStateMachine({ threshold: this.options.retry });
+    }
+
+    onStart() {
+        this.timer = window.setInterval(() => {
```

Comment:
Could it happen that there are several alive requests in parallel? Do we want it?

Reply:
> Could it happen that there are several alive requests in parallel?

Yes, it can. 

> Do we want it?

I do not think this will cause a problem but for the sake of simplicity, I change it.